src/renate/memory/buffer.py

```diff
@@ -197,189 +136,204 @@ def set_transforms(
-        self._class_to_index_map = defaultdict(list)
+    def state_dict(self) -> Dict:
+        state_dict = super().state_dict()
+        state_dict["indices_by_class"] = self._indices_by_class
+        return state_dict
 
-    def _update(self, dataset: Dataset, metadata: DataDict) -> None:
+    def load_state_dict(self, state_dict: Dict) -> None:
+        super().load_state_dict(state_dict)
+        self._indices_by_class = defaultdict(list)
+        self._indices_by_class.update(state_dict["indices_by_class"])
+
+    def _get_largest_class(self):
+        largest_classes = []
+        max_length = 0
+        for y, indices in self._indices_by_class.items():
+            if len(indices) > max_length:
+                largest_classes = [y]
+                max_length = len(indices)
+            if len(indices) == max_length:
+                largest_classes.append(y)
+        rand = torch.randint(low=0, high=len(largest_classes), size=(1,), generator=self._rng)
+        largest_class = largest_classes[rand.item()]
+        return largest_class, max_length
+
+    def _update(self, dataset: Dataset, metadata: Dict) -> Dict[int, int]:
+        assignments = {}
         for i in range(len(dataset)):
-            if len(self) < self._max_size:
-                self._append(dataset[i], {key: value[i] for key, value in metadata.items()})
-                self._record_class_index(dataset[i], len(self) - 1)
+            y = int(dataset[i][1])
+            if self._count < self._max_size:
+                assignments[self._count] = i
+                self._indices_by_class[y].append(self._count)
             else:
-                class_counts_max_value = max(
-                    [len(self._class_to_index_map[key]) for key in self._class_to_index_map.keys()]
-                )
-                largest_classes = [
-                    k
-                    for k, v in self._class_to_index_map.items()
-                    if len(v) == class_counts_max_value
-                ]
-                largest_class = largest_classes[
-                    torch.randint(
-                        low=0, high=len(largest_classes), size=(1,), generator=self._rng
-                    ).item()
-                ]
-                memory_idx_to_replace = self._class_to_index_map[largest_class][
-                    torch.randint(
-                        low=0,
-                        high=len(self._class_to_index_map[largest_class]),
-                        size=(1,),
-                        generator=self._rng,
-                    ).item()
-                ]
-                self[memory_idx_to_replace] = (
-                    dataset[i],
-                    {key: value[i] for key, value in metadata.items()},
-                )
-                self._class_to_index_map[largest_class].remove(memory_idx_to_replace)
-                self._record_class_index(dataset[i], memory_idx_to_replace)
-
+                largest_class, max_length = self._get_largest_class()
```

Comment:
not very efficient (compute and code). one could keep a counter around (array mapping class index to count) and then use argmax.

Reply:
Agreed, but let's fix separately, okay?